controllers/vaultstaticsecret_controller.go

```diff
@@ -218,11 +246,255 @@ func (r *VaultStaticSecretReconciler) handleDeletion(ctx context.Context, o clie
 	return nil
 }
 
+func (r *VaultStaticSecretReconciler) ensureEventWatcher(ctx context.Context, o *secretsv1beta1.VaultStaticSecret, c vault.Client) error {
+	logger := log.FromContext(ctx).WithName("ensureEventWatcher")
+	name := client.ObjectKeyFromObject(o)
+
+	meta, ok := r.eventWatcherRegistry.Get(name)
+	if ok {
+		// The watcher is running, and if the VSS object has not been updated,
+		// and the client ID is the same, just return
+		if meta.LastGeneration == o.GetGeneration() && meta.LastClientID == c.ID() {
+			logger.V(consts.LogLevelDebug).Info("Event watcher already running",
+				"namespace", o.Namespace, "name", o.Name)
+			return nil
+		}
+	}
+	if meta != nil {
+		// The watcher is running, but the metadata or vault client has changed,
+		// so kill it
+		if meta.Cancel != nil {
+			meta.Cancel()
+			// Wait for the goroutine to stop and remove itself from the event registry
+			waitCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+			defer cancel()
+			if err := waitForStoppedCh(waitCtx, meta.StoppedCh); err != nil {
+				logger.Error(err, "Failed to stop event watcher for VSS", "name", name)
+			}
+		} else {
+			logger.Error(fmt.Errorf("nil cancel function"), "event watcher has nil cancel function", "VSS", name, "meta", meta)
+		}
+	}
+	wsClient, err := c.WebsocketClient(kvEventPath)
+	if err != nil {
+		return fmt.Errorf("failed to create websocket client: %w", err)
+	}
+
+	watchCtx, cancel := context.WithCancel(context.Background())
+	stoppedCh := make(chan struct{}, 1)
+	updatedMeta := &eventWatcherMeta{
+		Cancel:         cancel,
+		LastClientID:   c.ID(),
+		LastGeneration: o.GetGeneration(),
+		StoppedCh:      stoppedCh,
+	}
+	// launch the goroutine to watch events
+	logger.V(consts.LogLevelDebug).Info("Starting event watcher", "meta", updatedMeta)
+	r.eventWatcherRegistry.Register(name, updatedMeta)
+	go r.getEvents(watchCtx, o, wsClient, stoppedCh)
+
+	return nil
+}
+
+// unWatchEvents - If the VSS is in the registry, cancel its event watcher
+// context to close the goroutine, and remove the VSS from the registry
+func (r *VaultStaticSecretReconciler) unWatchEvents(o *secretsv1beta1.VaultStaticSecret) {
+	name := client.ObjectKeyFromObject(o)
+	meta, ok := r.eventWatcherRegistry.Get(name)
+	if ok {
+		if meta.Cancel != nil {
+			meta.Cancel()
+		}
+		r.eventWatcherRegistry.Delete(name)
+	}
+}
+
+// getEvents calls streamStaticSecretEvents in a loop, collecting and responding
+// to any errors returned.
+func (r *VaultStaticSecretReconciler) getEvents(ctx context.Context, o *secretsv1beta1.VaultStaticSecret, wsClient *vault.WebsocketClient, stoppedCh chan struct{}) {
+	logger := log.FromContext(ctx).WithName("getEvents")
+	name := client.ObjectKeyFromObject(o)
+	defer func() {
+		r.eventWatcherRegistry.Delete(name)
+		close(stoppedCh)
+	}()
+
+	// Use the same backoff options used for Vault reads in Reconcile()
+	retryBackoff := backoff.NewExponentialBackOff(r.BackOffRegistry.opts...)
```

Comment:
probably should call `retryBackoff.Reset()`

Reply:
It looks like backoff.NewExponentialBackOff() already calls Reset()? https://github.com/cenkalti/backoff/blob/v4.3.0/exponential.go#L100